hamilton/function_modifiers/expanders.py

```diff
@@ -869,6 +874,156 @@ def extractor_fn(
+    :return: List of field types.
+    """
+
+    base_type = typing_inspect.get_origin(output_type)
+    if base_type != tuple and base_type != Tuple:
+        return
+
+    output_args = typing_inspect.get_args(output_type)
+    num_ellipsis = output_args.count(Ellipsis)
+    if num_ellipsis > 1:
+        raise base.InvalidDecoratorException(
+            f"Invalid tuple: Found more than one ellipsis ('...'): {output_type}"
+        )
+    elif num_ellipsis == 1:
+        if len(output_args) != 2:
+            raise base.InvalidDecoratorException(
+                f"Invalid tuple: Ellipsis ('...') must be second element: {output_type}"
+            )
+        # Valid Indeterminate length tuple, e.g. `tuple[int, ...]`, `typing.Tuple[int, ...]`
+        output_args = tuple(output_args[0] for _ in range(len(fields)))
+
+    if len(output_args) < len(fields):
+        raise base.InvalidDecoratorException(
+            f"Number of unpacked fields ({len(fields)}) is greater than the number of fields in "
+            f"the output type ({len(output_args)}): {output_type}"
+        )
+
+    errors = []
+    field_types = []
+    for idx, arg in enumerate(output_args):
+        # Determine if the type is a valid type. Note that for Python <3.11, `Any` is not a type
+        if not (
+            isinstance(arg, type)
+            or arg is Any
+            or typing_inspect.is_generic_type(arg)
+            or typing_inspect.is_union_type(arg)
+        ):
+            field_name = fields[idx]
+            errors.append(f"Field {field_name} (index {idx}) does not declare a valid type: {arg}")
+        field_types.append(arg)
+
+    if errors:
+        raise base.InvalidDecoratorException(f"Found errors in the output type: {errors}")
+
+    return field_types
+
+
+class unpack_fields(base.SingleNodeNodeTransformer):
+    """Extracts fields from a tuple output.
+
+    Expands a single function into the following nodes:
+
+    - 1 function that outputs the original tuple
+    - n functions, each of which take in the original tuple and output a specific field
+
+    The decorated function must have an return type of either `tuple` (python 3.9+) or
+    `typing.Tuple`, and must specify either:
+    - An explicit length tuple (e.g.`tuple[int, str]`, `typing.Tuple[int, str]`)
+    - An indeterminate length tuple (e.g. `tuple[int, ...]`, `typing.Tuple[int, ...]`)
+
+    :param fields: Fields to extract from the return value of the decorated function.
+    """
+
+    output_type: Any
+    field_types: List[Type]
+
+    def __init__(self, *fields: str):
+        super().__init__()
+        self.fields = list(fields)
+
+    @override
+    def validate(self, fn: Callable):
+        output_type = typing.get_type_hints(fn).get("return")
+        field_types = _process_unpack_fields(self.fields, output_type)
+        if field_types:
+            self.field_types = field_types
+            self.output_type = output_type
+        else:
+            message = (
```

Comment:
Feel like `_process_unpack_fields` should raise an error in this case? Is there a valid reason to have an empty return?

Reply:
You know what, originally I didn't want to pass `output_type` to `_process_unpack_fields` and this was my way of still producing a nice error message. However, now that I changed my mind about `output_type` it's definitely better to raise the error in `_process_unpack_fields` - it's cleaner that way.